src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,635 @@
+						errhint(
+							"First, add the coordinator with master_add_node command")));
+	}
+}
+
+
+/*
+ * ErrorIfUnsupportedCitusLocalTableKind errors out if the relation kind of
+ * relation with relationId is not supported for citus local table creation.
+ */
+static void
+ErrorIfUnsupportedCitusLocalTableKind(Oid relationId)
+{
+	const char *relationName = get_rel_name(relationId);
+
+	Assert(relationName != NULL);
+
+	if (IsChildTable(relationId) || IsParentTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", citus local "
+							   "tables cannot be involved in inheritance relationships",
+							   relationName)));
+	}
+
+	if (PartitionTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", citus local "
+							   "tables cannot be partition of other tables",
+							   relationName)));
+	}
+
+	char relationKind = get_rel_relkind(relationId);
+	if (!(relationKind == RELKIND_RELATION || relationKind == RELKIND_FOREIGN_TABLE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot create citus local table \"%s\", only regular "
+							   "tables and foreign tables are supported for citus local "
+							   "table creation", relationName)));
+	}
+}
+
+
+/*
+ * CreateCitusLocalTable is the internal method that creates a citus table
+ * from the table with relationId. The created table would have the following
+ * properties:
+ *  - it will have only one shard,
+ *  - its distribution method will be DISTRIBUTE_BY_NONE,
+ *  - its replication model will be ReplicationModel,
+ *  - its replication factor will be set to 1.
+ * Similar to reference tables, it has only 1 placement. In addition to that, that
+ * single placement is only allowed to be on the coordinator.
+ */
+static void
+CreateCitusLocalTable(Oid relationId)
+{
+	/* these checks should be done before acquiring the lock on the table */
+	EnsureCoordinator();
+	EnsureTableOwner(relationId);
+
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	ErrorIfUnsupportedCreateCitusLocalTable(relationId);
+
+	ObjectAddress tableAddress = { 0 };
+	ObjectAddressSet(tableAddress, RelationRelationId, relationId);
+
+	/*
+	 * Ensure dependencies first as we will create shell table on the other nodes
+	 * in the MX case.
+	 */
+	EnsureDependenciesExistOnAllNodes(&tableAddress);
+
+	/*
+	 * Make sure that existing reference tables have been replicated to all
+	 * the nodes such that we can create foreign keys and joins work
+	 * immediately after creation.
+	 */
+	EnsureReferenceTablesExistOnAllNodes();
+
+	List *shellTableDDLEvents = GetShellTableDDLEventsForCitusLocalTable(relationId);
```

Comment:
Thanks for using blocks here, I think we can further use more blocks by separating the part until this like into another function and after the call InsertMetadataForCitusLocalTable to another function as pre create table operations and post create table operations. I think there might be common operations between different table types and we might need to call the methods from those places as well, but still I suggest creating these methods at least for citus local tables in this PR so that this function is even in better form and later in a separate PR we can use the methods(with small changes) in other places as well.